File: project/src/nwr/gui/SkillsWindow.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package nwr.gui;

import jzrlib.core.StringList;
import nwr.core.Locale;
import nwr.core.RS;
import nwr.core.StaticData;
import nwr.core.types.CreatureAction;
import nwr.core.types.PlaceFlags;
import nwr.core.types.PlaceID;
import nwr.core.types.PlaceRec;
import nwr.core.types.SkillID;
import nwr.core.types.SkillKind;
import nwr.database.CreatureEntry;
import nwr.database.ItemEntry;
import nwr.database.ItemEntry.EffectEntry;
import nwr.effects.EffectID;
import nwr.effects.EffectParams;
import nwr.effects.EffectRec;
import nwr.effects.EffectsData;
import nwr.engine.BaseControl;
import nwr.engine.BaseScreen;
import nwr.engine.KeyEventArgs;
import nwr.engine.Keys;
import nwr.engine.MouseButton;
import nwr.engine.WindowStyles;
import nwr.gui.controls.CtlCommon;
import nwr.gui.controls.LBItem;
import nwr.gui.controls.LBOptions;
import nwr.gui.controls.ListBox;
import nwr.main.GlobalVars;
import nwr.player.Faith;
import nwr.player.Player;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class SkillsWindow extends NWWindow
{
    public static final int SWM_SKILLS = 0;
    public static final int SWM_SPELLS = 1;
    public static final int SWM_TILES = 2;
    public static final int SWM_LANDS = 3;
    public static final int SWM_SCROLLS = 4;
    public static final int SWM_GODS = 5;
    
    private int fMode;
    private final ListBox fSkillsList;

    public SkillsWindow(BaseControl owner)
    {
        super(owner);
        super.setFont(CtlCommon.smFont);
        super.setWidth(600);
        super.setHeight(460);
        super.WindowStyle = new WindowStyles(WindowStyles.wsModal, WindowStyles.wsKeyPreview);
        super.Shifted = true;
        super.BackDraw = false;

        this.fSkillsList = new ListBox(this);
        this.fSkillsList.setLeft(10);
        this.fSkillsList.setTop(10);
        this.fSkillsList.setWidth(super.getWidth() - 20);
        this.fSkillsList.setHeight(super.getHeight() - 20);
        this.fSkillsList.setItemHeight(34);
        this.fSkillsList.setIconHeight(30);
        this.fSkillsList.setIconWidth(32);
        this.fSkillsList.Options.include(LBOptions.lboIcons);
        this.fSkillsList.setVisible(true);
        this.fSkillsList.setColumns(2);
        this.fSkillsList.OnItemSelect = this::onSkillsSelect;
        this.fSkillsList.OnKeyDown = this::onSkillsKeyDown;
    }

    @Override
    protected void dispose(boolean disposing)
    {
        if (disposing) {
            this.fSkillsList.dispose();
        }
        super.dispose(disposing);
    }

    private void onSkillsKeyDown(Object Sender, KeyEventArgs eventArgs)
    {
        if (eventArgs.Key == Keys.GK_RETURN && this.fSkillsList.getSelIndex() >= 0) {
            this.onSkillsSelect(Sender, MouseButton.mbRight, this.fSkillsList.getItems().getItem(this.fSkillsList.getSelIndex()));
        }
    }

    private void onSkillsSelect(Object Sender, MouseButton Button, LBItem Item)
    {
        if (Button == MouseButton.mbRight) {
            int idx = (int) Item.Data;
            switch (this.fMode) {
                case SWM_SKILLS: {
                    SkillID sk = SkillID.forValue(idx);
                    if (sk != SkillID.Sk_Alchemy) {
                        if (sk != SkillID.Sk_Ironworking) {
                            if (sk != SkillID.Sk_Spellcasting) {
                                this.hide();
                                GlobalVars.nwrGame.doPlayerAction(CreatureAction.caSkillUse, idx);
                            } else {
                                this.setMode(SWM_SPELLS);
                            }
                        } else {
                            String ts = GlobalVars.nwrGame.getPlayer().getCraft().checkSmithyTools();
                            if (ts.compareTo("") != 0) {
                                GlobalVars.nwrWin.showText(GlobalVars.nwrWin, Locale.format(RS.rs_NoInstruments, new Object[]{ts}));
                            } else {
                                this.hide();
                                GlobalVars.nwrWin.showSmithyWin();
                            }
                        }
                    } else {
                        this.hide();
                        GlobalVars.nwrWin.showAlchemyWin();
                    }
                    break;
                }

                case SWM_SPELLS:
                    this.hide();
                    GlobalVars.nwrGame.doPlayerAction(CreatureAction.caSpellUse, idx);
                    break;

                case SWM_TILES:
                    this.hide();
                    GlobalVars.nwrWin.getTargetData().Ext.setParam(EffectParams.ep_TileID, idx);
                    GlobalVars.nwrWin.useTarget();
                    break;

                case SWM_LANDS:
                    this.hide();
                    GlobalVars.nwrWin.getTargetData().Ext.setParam(EffectParams.ep_Land, idx);
                    GlobalVars.nwrWin.useTarget();
                    break;

                case SWM_SCROLLS:
                    this.hide();
                    GlobalVars.nwrWin.getTargetData().Ext.setParam(EffectParams.ep_ScrollID, idx);
                    GlobalVars.nwrWin.useTarget();
                    break;

                case SWM_GODS:
                    this.hide();
                    GlobalVars.nwrWin.getTargetData().Ext.setParam(EffectParams.ep_GodID, idx);
                    GlobalVars.nwrWin.useTarget();
                    break;
            }
        }
    }

    public final void setMode(int value)
    {
        this.fMode = value;
        this.fSkillsList.getItems().clear();
        Player player = GlobalVars.nwrGame.getPlayer();

        switch (this.fMode) {
            case SWM_SKILLS: {
                GlobalVars.nwrWin.showTextAux(Locale.getStr(RS.rs_UseWhichSkill));

                for (int sk = SkillID.Sk_First; sk <= SkillID.Sk_Last; sk++) {
                    SkillID sid = SkillID.forValue(sk);
                    if (player.getSkill(sid) > 0) {
                        String s = Locale.getStr(sid.NameRS) + " (" + Locale.getStr(RS.rs_Level) + " " + String.valueOf(player.getSkill(sid)) + ")";
                        int imageIndex;
                        if (sid.Kinds == SkillKind.ssInnatePower) {
                            imageIndex = sid.ImageIndex;
                        } else {
                            imageIndex = sid.ImageIndex;
                        }
                        this.addItem(s, (Object) (int) sk, imageIndex);
                    }
                }
                break;
            }

            case SWM_SPELLS: {
                GlobalVars.nwrWin.showTextAux(Locale.getStr(RS.rs_CastWhichSpell));

                int skill_Spellcasting = player.getSkill(SkillID.Sk_Spellcasting);
                for (int eff = EffectID.eid_First; eff <= EffectID.eid_Last; eff++) {
                    EffectRec efRec = EffectsData.dbEffects[eff];
                    if (skill_Spellcasting >= (int) efRec.LevReq && efRec.MPReq <= player.MPCur) {
                        String st = " (" + String.valueOf(efRec.MPReq) + ")";
                        this.addItem(Locale.getStr(efRec.NameRS) + st, eff, SkillID.Sk_Spellcasting.ImageIndex);
                    }
                }

                break;
            }

            case SWM_TILES: {
                // "Change the terrain where?"
                GlobalVars.nwrWin.showTextAux(Locale.getStr(RS.rs_CreateWhichTerrain));

                for (int pd = PlaceID.pid_First; pd <= PlaceID.pid_Last; pd++) {
                    PlaceRec pRec = StaticData.dbPlaces[pd];
                    if ((pRec.Signs.contains(PlaceFlags.psCanCreate))) {
                        this.addItem(Locale.getStr(pRec.NameRS), pd, pRec.ImageIndex);
                    }
                }

                break;
            }

            case SWM_LANDS: {
                GlobalVars.nwrWin.showTextAux(Locale.getStr(RS.rs_TravelWhere));
                StringList lst = GlobalVars.nwrGame.getVisitedLands();

                int num = lst.getCount();
                for (int i = 0; i < num; i++) {
                    this.addItem(lst.get(i), lst.getObject(i), -1);
                }
                lst.dispose();
                break;
            }

            case SWM_SCROLLS: {
                // "Rewrite which?"
                GlobalVars.nwrWin.showTextAux(Locale.getStr(RS.rs_WriteWhichScroll));

                int num2 = GlobalVars.dbScrolls.getCount();
                for (int i = 0; i < num2; i++) {
                    ItemEntry scr = (ItemEntry) GlobalVars.nwrBase.getEntry(GlobalVars.dbScrolls.get(i));
                    EffectEntry[] effects = scr.Effects;
                    if (((effects != null) ? effects.length : 0) >= 1) {
                        EffectID eff = scr.Effects[0].EffID;
                        if (player.getSkill(SkillID.Sk_Spellcasting) >= (int) EffectsData.dbEffects[eff.getValue()].LevReq) {
                            this.addItem(scr.getName(), GlobalVars.dbScrolls.get(i), -1);
                        }
                    }
                }
                break;
            }

            case SWM_GODS:
                for (int i = 0; i < Faith.Pantheon.length; i++) {
                    CreatureEntry deity = (CreatureEntry) GlobalVars.nwrBase.findEntryBySign(Faith.Pantheon[i].Sign);
                    if (deity != null) {
                        this.addItem(deity.getName(), deity.GUID, deity.ImageIndex);
                    }
                }
                break;
        }
    }

    private void addItem(String aName, Object aObj, int aImageIndex)
    {
        LBItem listItem = this.fSkillsList.getItems().add(aName, aObj);
        listItem.Color = BaseScreen.clGold;
        listItem.ImageIndex = aImageIndex;
    }

    @Override
    protected void doShowEvent()
    {
        super.doShowEvent();
        this.fSkillsList.ImagesList = GlobalVars.nwrWin.Resources;
        super.setActiveControl(this.fSkillsList);
    }
}

File: project/GUI/SkillsWindow.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih.
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using BSLib;
using NWR.Core;
using NWR.Core.Types;
using NWR.Database;
using NWR.Effects;
using NWR.Game;
using NWR.GUI.Controls;
using ZRLib.Core;
using ZRLib.Engine;

namespace NWR.GUI
{
    public sealed class SkillsWindow : NWWindow
    {
        public const int SWM_SKILLS = 0;
        public const int SWM_SPELLS = 1;
        public const int SWM_TILES = 2;
        public const int SWM_LANDS = 3;
        public const int SWM_SCROLLS = 4;
        public const int SWM_GODS = 5;

        private int fMode;
        private readonly ListBox fSkillsList;

        public SkillsWindow(BaseControl owner)
            : base(owner)
        {
            Font = CtlCommon.SmFont;
            Width = 600;
            Height = 460;
            WindowStyle = new WindowStyles(WindowStyles.wsModal, WindowStyles.wsKeyPreview);
            Shifted = true;
            BackDraw = false;

            fSkillsList = new ListBox(this);
            fSkillsList.Left = 10;
            fSkillsList.Top = 10;
            fSkillsList.Width = Width - 20;
            fSkillsList.Height = Height - 20;
            fSkillsList.ItemHeight = 34;
            fSkillsList.IconHeight = 30;
            fSkillsList.IconWidth = 32;
            fSkillsList.Options.Include(LBOptions.lboIcons);
            fSkillsList.Visible = true;
            fSkillsList.Columns = 2;
            fSkillsList.OnItemSelect = OnSkillsSelect;
            fSkillsList.OnKeyDown = OnSkillsKeyDown;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing) {
                fSkillsList.Dispose();
            }
            base.Dispose(disposing);
        }

        private void OnSkillsKeyDown(object Sender, KeyEventArgs eventArgs)
        {
            if (eventArgs.Key == Keys.GK_RETURN && fSkillsList.SelIndex >= 0) {
                OnSkillsSelect(Sender, MouseButton.mbRight, fSkillsList.Items.GetItem(fSkillsList.SelIndex));
            }
        }

        private void OnSkillsSelect(object Sender, MouseButton Button, LBItem Item)
        {
            if (Button == MouseButton.mbRight) {
                int idx = (int)Item.Data;
                switch (fMode) {
                    case SWM_SKILLS:
                        {
                            SkillID sk = (SkillID)(idx);
                            if (sk != SkillID.Sk_Alchemy) {
                                if (sk != SkillID.Sk_Ironworking) {
                                    if (sk != SkillID.Sk_Spellcasting) {
                                        Hide();
                                        GlobalVars.nwrGame.DoPlayerAction(CreatureAction.caSkillUse, idx);
                                    } else {
                                        Mode = SWM_SPELLS;
                                    }
                                } else {
                                    string ts = GlobalVars.nwrGame.Player.Craft.CheckSmithyTools();
                                    if (ts.CompareTo("") != 0) {
                                        GlobalVars.nwrWin.ShowText(GlobalVars.nwrWin, BaseLocale.Format(RS.rs_NoInstruments, new object[]{ ts }));
                                    } else {
                                        Hide();
                                        GlobalVars.nwrWin.ShowSmithyWin();
                                    }
                                }
                            } else {
                                Hide();
                                GlobalVars.nwrWin.ShowAlchemyWin();
                            }
                            break;
                        }

                    case SWM_SPELLS:
                        Hide();
                        GlobalVars.nwrGame.DoPlayerAction(CreatureAction.caSpellUse, idx);
                        break;

                    case SWM_TILES:
                        Hide();
                        GlobalVars.nwrWin.TargetData.Ext.SetParam(EffectParams.ep_TileID, idx);
                        GlobalVars.nwrWin.UseTarget();
                        break;

                    case SWM_LANDS:
                        Hide();
                        GlobalVars.nwrWin.TargetData.Ext.SetParam(EffectParams.ep_Land, idx);
                        GlobalVars.nwrWin.UseTarget();
                        break;

                    case SWM_SCROLLS:
                        Hide();
                        GlobalVars.nwrWin.TargetData.Ext.SetParam(EffectParams.ep_ScrollID, idx);
                        GlobalVars.nwrWin.UseTarget();
                        break;

                    case SWM_GODS:
                        Hide();
                        GlobalVars.nwrWin.TargetData.Ext.SetParam(EffectParams.ep_GodID, idx);
                        GlobalVars.nwrWin.UseTarget();
                        break;
                }
            }
        }

        public int Mode
        {
            set {
                fMode = value;
                fSkillsList.Items.Clear();
                Player player = GlobalVars.nwrGame.Player;
    
                switch (fMode) {
                    case SWM_SKILLS:
                        {
                            GlobalVars.nwrWin.ShowTextAux(BaseLocale.GetStr(RS.rs_UseWhichSkill));
    
                            for (var sk = SkillID.Sk_First; sk <= SkillID.Sk_Last; sk++) {
                                SkillRec skRec = StaticData.dbSkills[(int)sk];
                                if (player.GetSkill(sk) > 0) {
                                    string s = BaseLocale.GetStr(skRec.Name) + " (" + BaseLocale.GetStr(RS.rs_Level) + " " + Convert.ToString(player.GetSkill(sk)) + ")";
                                    int imageIndex;
                                    if (skRec.Kinds == SkillKind.ssInnatePower) {
                                        imageIndex = skRec.ImageIndex;
                                    } else {
                                        imageIndex = skRec.ImageIndex;
                                    }
                                    AddItem(s, (object)(int)sk, imageIndex);
                                }
                            }
                            break;
                        }
    
                    case SWM_SPELLS:
                        {
                            GlobalVars.nwrWin.ShowTextAux(BaseLocale.GetStr(RS.rs_CastWhichSpell));
    
                            int skill_Spellcasting = player.GetSkill(SkillID.Sk_Spellcasting);
                            for (var eff = EffectID.eid_First; eff <= EffectID.eid_Last; eff++) {
                                EffectRec efRec = EffectsData.dbEffects[(int)eff];
                                if (skill_Spellcasting >= (int)efRec.LevReq && efRec.MPReq <= player.MPCur) {
                                    string st = " (" + Convert.ToString(efRec.MPReq) + ")";
                                    AddItem(BaseLocale.GetStr(efRec.NameRS) + st, eff,
                                                 StaticData.dbSkills[(int)SkillID.Sk_Spellcasting].ImageIndex);
                                }
                            }
    
                            break;
                        }
    
                    case SWM_TILES:
                        {
                            // "Change the terrain where?"
                            GlobalVars.nwrWin.ShowTextAux(BaseLocale.GetStr(RS.rs_CreateWhichTerrain));
    
                            for (int pd = PlaceID.pid_First; pd <= PlaceID.pid_Last; pd++) {
                                PlaceRec pRec = StaticData.dbPlaces[pd];
                                if ((pRec.Signs.Contains(PlaceFlags.psCanCreate))) {
                                    AddItem(BaseLocale.GetStr(pRec.NameRS), pd, pRec.ImageIndex);
                                }
                            }
    
                            break;
                        }
    
                    case SWM_LANDS:
                        {
                            GlobalVars.nwrWin.ShowTextAux(BaseLocale.GetStr(RS.rs_TravelWhere));
                            StringList lst = GlobalVars.nwrGame.VisitedLands;
    
                            int num = lst.Count;
                            for (int i = 0; i < num; i++) {
                                AddItem(lst[i], lst.GetObject(i), -1);
                            }
                            lst.Dispose();
                            break;
                        }
    
                    case SWM_SCROLLS:
                        {
                            // "Rewrite which?"
                            GlobalVars.nwrWin.ShowTextAux(BaseLocale.GetStr(RS.rs_WriteWhichScroll));
    
                            int num2 = GlobalVars.dbScrolls.Count;
                            for (int i = 0; i < num2; i++) {
                                ItemEntry scr = (ItemEntry)GlobalVars.nwrDB.GetEntry(GlobalVars.dbScrolls[i]);
                                ItemEntry.EffectEntry[] effects = scr.Effects;
                                if (((effects != null) ? effects.Length : 0) >= 1) {
                                    EffectID eff = scr.Effects[0].EffID;
                                    if (player.GetSkill(SkillID.Sk_Spellcasting) >= (int)EffectsData.dbEffects[(int)eff].LevReq) {
                                        AddItem(scr.Name, GlobalVars.dbScrolls[i], -1);
                                    }
                                }
                            }
                            break;
                        }
    
                    case SWM_GODS:
                        for (int i = 0; i < Faith.Pantheon.Length; i++) {
                            CreatureEntry deity = (CreatureEntry)GlobalVars.nwrDB.FindEntryBySign(Faith.Pantheon[i].Sign);
                            if (deity != null) {
                                AddItem(deity.Name, deity.GUID, deity.ImageIndex);
                            }
                        }
                        break;
                }
            }
        }

        private void AddItem(string aName, object aObj, int aImageIndex)
        {
            LBItem listItem = fSkillsList.Items.Add(aName, aObj);
            listItem.Color = Colors.Gold;
            listItem.ImageIndex = aImageIndex;
        }

        protected override void DoShowEvent()
        {
            base.DoShowEvent();
            fSkillsList.ImagesList = GlobalVars.nwrWin.Resources;
            ActiveControl = fSkillsList;
        }
    }

}